### PR TITLE
CBSP-2990: Automatically Provision SSL Certs

### PR DESCRIPTION
**Pull request by orthogonal3**
Initial working version of automatic SSL certificate provisioning script for puppetized linux vagrants

---

### Generating SSL Certificates

SSL certificates for any node can be generated at provisioning time by toggling on the `VAGRANT_GENSSL` env var:

    VAGRANT_GENSSL=1 vagrant up

By default 2 client certificates are also created once for `cbssl-user1@cb.local` and `cbssl-user2@cb.local`, together with any certificates required for new vagrant nodes.

The server certificates will have SAN addresses for all DNS names, inc FQDN, and all IP addresses.

These will be generated in a "node_ssl" folder, with the following structure:

    node_ssl
    ├── ca
    │   ├── ca.key
    │   ├── ca.pem
    │   └── serial.srl
    ├── config.cnf
    ├── node
    │   ├── node1-shortname
    │   │   ├── chain.pem
    │   │   ├── node1-shortname.csr
    │   │   └── pkey.key
    │   ├── node2-shortname
    │   │   └── ... 3 files
    │   ├── node3-shortname
    │   │   └── ... 3 files
    │   └── node4-shortname
    │       └── ... 3 files
    └── user
        ├── cbssl-user1@cb.local
        │   ├── cbssl-user1@cb.local.csr
        │   ├── cbssl-user1@cb.local.key
        │   └── cbssl-user1@cb.local.pem
        └── cbssl-user2@cb.local
            └── ... 3 files

During generation, the node pkey.key, chain.pem, and ca.pem file will be copied to the Couchbase Server `inbox` directory, for easy use later.

#### Configuration

Once the cluster is online, the certificates can then be imported with the usual method:

First load the CA certificate on one node.
Warnings might appear that the old default node certifcates aren't signed by the new CA, this is fine.

From your local machine at the vagrant os-level dir (eg: `vagrants/6.5.0/centos7/`):

    curl -u Administrator -p -X POST --data-binary "@./node_ssl/ca/ca.pem" \
    http://<NODE ADDRESS>:8091/controller/uploadClusterCA

Then once per-node, reload the new certificate:

    curl -u Administrator -p -X POST \
    http://<NODE ADDRESS>:8091/node/controller/reloadCertificate

#### Testing

If desired, a server connection can be tested with `openssl` as usual:

    openssl s_client -connect <NODE ADDRESS>:18091 -CAfile node_ssl/ca/ca.pem <<< ""
    openssl s_client -connect <NODE ADDRESS>:11207 -CAfile node_ssl/ca/ca.pem <<< ""

Just look for `Verify return code: 0 (ok)` towards the end of the output to show the chain is acceptable, given the cluster CA certificate.

#### Configuration

An RBAC mapping must be made from the user certificate `Subject.CN` field to Couchbase Server username.

The default client certificate `Subject.CN` are `cbssl-user1@cb.local` and `cbssl-user2@cb.local`.
This format should be easy to parse in Couchbase Server, allowing different options.

For example:

* Prefix: `cbssl-`
* Delimiter: `@`

Yields usernames of `user1` and `user2`.
Ensure to set the mapping to enabled, and not mandatory unless wanting to authenticate ALL requests with a certificate.

#### Testing

The client certificate can be tested against the cluster with `cbc-pillowfight`.
From your local machine at the vagrant os-level dir (eg: `vagrants/6.5.0/centos7/`):

    cbc-pillowfight -U "couchbases://<NODE ADDRESS>/<BUCKET NAME>" \
    --truststorepath="node_ssl/ca/ca.pem" \
    --keypath="node_ssl/user/cbssl-user1@cb.local/cbssl-user1@cb.local.key" \
    --certpath="node_ssl/user/cbssl-user1@cb.local/cbssl-user1@cb.local.pem"

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -69,6 +71,103 @@ If the user has the *landrush* plugin installed, the TLD will be set to `.vagran
...
 
+## SSL Certificates
+
+### Generating SSL Certificates
+
+SSL certificates for any node can be generated at provisioning time by toggling on the `VAGRANT_GENSSL` env var:
+
+    VAGRANT_GENSSL=1 vagrant up
+
+By default 2 client certificates are also created once for `cbssl-user1@cb.local` and `cbssl-user2@cb.local`, together with any certificates required for new vagrant nodes.
```
Comment (Tim020) [inline README.md]: “created once” -> “created, one” (?)

Comment (orthogonal3) [reply README.md]: I was more meaning if you start 4 nodes, the user certs are only created once, when the first server boots. Second server just skips that bit.

Comment (Tim020) [reply README.md]: Ahh, I see! Maybe worth clarifying that then :)

**Comment by orthogonal3** [comment]
New version written with internal root CA handling, discarding this PR - will reopen with new one.

**Pull request closed without merging**